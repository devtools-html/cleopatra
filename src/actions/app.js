/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { getSelectedTab, getDataSource } from '../selectors/url-state';
import { getTrackThreadHeights } from '../selectors/app';
import { sendAnalytics } from '../utils/analytics';
import type { ThreadIndex } from '../types/profile';
import type { CssPixels } from '../types/units';
import type { Action, ThunkAction } from '../types/store';
import type { TabSlug } from '../app-logic/tabs-handling';
import type { UrlState } from '../types/state';

export function changeSelectedTab(selectedTab: TabSlug): ThunkAction<void> {
  return (dispatch, getState) => {
    const previousTab = getSelectedTab(getState());
    if (previousTab !== selectedTab) {
      sendAnalytics({
        hitType: 'pageview',
        page: selectedTab,
      });
      dispatch({
        type: 'CHANGE_SELECTED_TAB',
        selectedTab,
      });
    }
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeProfilesToCompare(profiles: string[]): Action {
  return {
    type: 'CHANGE_PROFILES_TO_COMPARE',
    profiles,
  };
}

export function urlSetupDone(): ThunkAction<void> {
  return (dispatch, getState) => {
    dispatch({ type: 'URL_SETUP_DONE' });

    // After the url setup is done, we can successfully query our state about its
    // initial page.
    const dataSource = getDataSource(getState());
    sendAnalytics({
      hitType: 'pageview',
      page: dataSource === 'none' ? 'home' : getSelectedTab(getState()),
    });
    sendAnalytics({
      hitType: 'event',
      eventCategory: 'datasource',
      eventAction: dataSource,
    });
  };
}

export function show404(url: string): Action {
  return { type: 'ROUTE_NOT_FOUND', url };
}

export function changeSidebarOpenState(tab: TabSlug, isOpen: boolean): Action {
  return { type: 'CHANGE_SIDEBAR_OPEN_STATE', tab, isOpen };
}

export function invalidatePanelLayout(): Action {
  return { type: 'INCREMENT_PANEL_LAYOUT_GENERATION' };
}

/**
 * The viewport component provides a hint to use shift to zoom scroll. The first
 * time a user does this, the hint goes away.
 */
export function setHasZoomedViaMousewheel() {
  return { type: 'HAS_ZOOMED_VIA_MOUSEWHEEL' };
}

/**
 * This function is called when a browser navigation event happens. A new UrlState
 * is generated when the window.location is serialized, or the state is pulled out of
 * the history API.
 */
export function updateUrlState(newUrlState: UrlState | null): Action {
  return { type: 'UPDATE_URL_STATE', newUrlState };
}

export const reportTrackThreadHeight = (
  threadIndex: ThreadIndex,
  height: CssPixels
): ThunkAction<void> => (dispatch, getState) => {
  const trackThreadHeights = getTrackThreadHeights(getState());
  const previousHeight = trackThreadHeights[threadIndex];
  if (previousHeight !== height) {
    // Guard against unnecessary dispatches. This could happen frequently.
    dispatch({
      type: 'UPDATE_TRACK_THREAD_HEIGHT',
      height,
      threadIndex,
    });
  }
};

/**
 * This action dismisses the newly published state. This happens when a user first
 * uploads a profile. We only want to remember this when we fist open the profile.
 */
export function dismissNewlyPublished(): Action {
  return { type: 'DISMISS_NEWLY_PUBLISHED' };
}
