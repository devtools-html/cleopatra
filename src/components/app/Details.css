.Details {
  display: flex;
  flex: auto;
  flex-direction: column;
}

.sidebar-open-close-button {
  width: 24px;
  height: 24px;
  flex: none;
  margin: 0 4px;
  background-position: center;
  background-repeat: no-repeat;
}

.sidebar-open-close-button-isopen {
  background-image: url(../../../res/img/svg/pane-collapse.svg);
}

.sidebar-open-close-button-isclosed {
  background-image: url(../../../res/img/svg/pane-expand.svg);
}
