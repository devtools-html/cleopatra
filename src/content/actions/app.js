// @flow
import type { Action } from './types';

export function changeSelectedTab(selectedTab: string): Action {
  return {
    type: 'CHANGE_SELECTED_TAB',
    selectedTab,
  };
}

export function profilePublished(hash: string): Action {
  return {
    type: 'PROFILE_PUBLISHED',
    hash,
  };
}

export function changeTabOrder(tabOrder: number[]): Action {
  return {
    type: 'CHANGE_TAB_ORDER',
    tabOrder,
  };
}
