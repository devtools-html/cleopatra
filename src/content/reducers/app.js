/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';

import type { Action } from '../actions/types';
import type { State, AppState, AppViewState, Reducer } from './types';

function view(state: AppViewState = { phase: 'INITIALIZING' }, action: Action): AppViewState {
  switch (action.type) {
    case 'TEMPORARY_ERROR_RECEIVING_PROFILE_FROM_WEB':
      return {
        phase: 'INITIALIZING',
        additionalData: { attempt: action.error.attempt },
      };
    case 'ERROR_RECEIVING_PROFILE_FROM_FILE':
    case 'ERROR_RECEIVING_PROFILE_FROM_ADDON':
    case 'FATAL_ERROR_RECEIVING_PROFILE_FROM_WEB':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_PROFILE_FROM_ADDON':
    case 'RECEIVE_PROFILE_FROM_WEB':
    case 'RECEIVE_PROFILE_FROM_FILE':
      return { phase: 'PROFILE' };
    default:
      return state;
  }
}

function isURLSetupDone(state: boolean = false, action: Action) {
  switch (action.type) {
    case '@@urlenhancer/urlSetupDone':
      return true;
    default:
      return state;
  }
}
const appStateReducer: Reducer<AppState> = combineReducers({ view, isURLSetupDone });
export default appStateReducer;

export const getApp = (state: State): AppState => state.app;
export const getView = (state: State): AppViewState => getApp(state).view;
export const getIsURLSetupDone = (state: State): boolean => getApp(state).isURLSetupDone;
