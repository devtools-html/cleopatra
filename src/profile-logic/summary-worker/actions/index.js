/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

import { summarizeProfile } from '../../../profile-logic/summarize-profile';

export function processProfileSummary() {
  return function (dispatch, getState) {
    dispatch({
      toContent: true,
      type: 'PROFILE_SUMMARY_PROCESSED',
      summary: summarizeProfile(getState().profile),
    });
  };
}

export function profileProcessed(profile) {
  return {
    type: 'PROFILE_PROCESSED',
    profile: profile,
  };
}
