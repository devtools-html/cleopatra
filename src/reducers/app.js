/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import { combineReducers } from 'redux';
import { tabSlugs } from '../app-logic/tabs-handling';

import type { TabSlug } from '../app-logic/tabs-handling';
import type {
  AppState,
  AppViewState,
  IsSidebarOpenPerPanelState,
  Reducer,
} from '../types/state';

const view: Reducer<AppViewState> = (
  state = { phase: 'INITIALIZING' },
  action
) => {
  if (state.phase === 'DATA_LOADED') {
    // Let's not come back at another phase if we're already displaying a profile
    return state;
  }

  switch (action.type) {
    case 'TEMPORARY_ERROR':
      return {
        phase: 'INITIALIZING',
        additionalData: {
          message: action.error.message,
          attempt: action.error.attempt,
        },
      };
    case 'FATAL_ERROR':
      return { phase: 'FATAL_ERROR', error: action.error };
    case 'WAITING_FOR_PROFILE_FROM_ADDON':
      return { phase: 'INITIALIZING' };
    case 'ROUTE_NOT_FOUND':
      return { phase: 'ROUTE_NOT_FOUND' };
    case 'RECEIVE_ZIP_FILE':
    case 'VIEW_PROFILE':
      return { phase: 'DATA_LOADED' };
    default:
      return state;
  }
};

const isUrlSetupDone: Reducer<boolean> = (state = false, action) => {
  switch (action.type) {
    case 'URL_SETUP_DONE':
      return true;
    default:
      return state;
  }
};

const hasZoomedViaMousewheel: Reducer<boolean> = (state = false, action) => {
  switch (action.type) {
    case 'HAS_ZOOMED_VIA_MOUSEWHEEL': {
      return true;
    }
    default:
      return state;
  }
};

function _getNoSidebarsOpen() {
  const state = {};
  tabSlugs.forEach(tabSlug => (state[tabSlug] = false));
  return state;
}

const isSidebarOpenPerPanel: Reducer<IsSidebarOpenPerPanelState> = (
  state = _getNoSidebarsOpen(),
  action
) => {
  switch (action.type) {
    case 'CHANGE_SIDEBAR_OPEN_STATE': {
      const { tab, isOpen } = action;
      // Due to how this action will be dispatched we'll always have the value
      // changed so we don't need the performance optimization of checking the
      // stored value against the new value.
      return {
        ...state,
        [tab]: isOpen,
      };
    }
    default:
      return state;
  }
};

/**
 * The panels that make up the timeline, details view, and sidebar can all change
 * their sizes depending on the state that is fed to them. In order to control
 * the invalidations of this sizing information, provide a "generation" value that
 * increases monotonically for any change that potentially changes the sizing of
 * any of the panels. This provides a mechanism for subscribing components to
 * deterministically update their sizing correctly.
 */
const panelLayoutGeneration: Reducer<number> = (state = 0, action) => {
  switch (action.type) {
    case 'INCREMENT_PANEL_LAYOUT_GENERATION':
    // Sidebar: (fallthrough)
    case 'CHANGE_SIDEBAR_OPEN_STATE':
    // Timeline: (fallthrough)
    case 'HIDE_GLOBAL_TRACK':
    case 'SHOW_GLOBAL_TRACK':
    case 'ISOLATE_PROCESS':
    case 'ISOLATE_PROCESS_MAIN_THREAD':
    case 'HIDE_LOCAL_TRACK':
    case 'SHOW_LOCAL_TRACK':
    case 'ISOLATE_LOCAL_TRACK':
    // Committed range changes: (fallthrough)
    case 'COMMIT_RANGE':
    case 'POP_COMMITTED_RANGES':
      return state + 1;
    default:
      return state;
  }
};

/**
 * Clicking on tracks can switch between different tabs. This piece of state holds
 * on to the last relevant thread-based tab that was viewed. This makes the UX nicer
 * for when a user clicks on a Network track, and gets taken to the Network
 * panel, then clicks on a thread or process track. With this state we can smoothly
 * transition them back to the panel they were using.
 */
const lastVisibleThreadTabSlug: Reducer<TabSlug> = (
  state = 'calltree',
  action
) => {
  switch (action.type) {
    case 'SELECT_TRACK':
    case 'CHANGE_SELECTED_TAB':
      if (action.selectedTab !== 'network-chart') {
        return action.selectedTab;
      }
      return state;
    default:
      return state;
  }
};

const appStateReducer: Reducer<AppState> = combineReducers({
  view,
  isUrlSetupDone,
  hasZoomedViaMousewheel,
  isSidebarOpenPerPanel,
  panelLayoutGeneration,
  lastVisibleThreadTabSlug,
});

export default appStateReducer;
