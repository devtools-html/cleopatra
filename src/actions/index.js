/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as app from './app';
import * as icons from './icons';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';
import * as profileUpload from './profile-upload';
import * as shortUrl from './short-url';
import * as urlState from './url-state';

export default Object.assign(
  {},
  app,
  icons,
  profileSummary,
  profileUpload,
  profileView,
  receiveProfile,
  shortUrl,
  timeline,
  urlState
);
