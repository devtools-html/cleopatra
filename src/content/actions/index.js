// @flow
import * as app from './app';
import * as profileSummary from './profile-summary';
import * as profileView from './profile-view';
import * as receiveProfile from './receive-profile';
import * as timeline from './timeline';

export default Object.assign({}, app, profileSummary, profileView, receiveProfile, timeline);
