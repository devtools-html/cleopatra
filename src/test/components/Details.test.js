/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import * as React from 'react';
import { shallowWithStore } from '../fixtures/enzyme';

import Details from '../../components/app/Details';
import { changeSelectedTab, changeSidebarOpenState } from '../../actions/app';
import { storeWithProfile } from '../fixtures/stores';
import { getProfileFromTextSamples } from '../fixtures/profiles/make-profile';

import { tabSlugs } from '../../app-logic/tabs-handling';
import type { TabSlug } from '../../app-logic/tabs-handling';

describe('app/Details', function() {
  const { profile } = getProfileFromTextSamples(`
    A A A
    B B B
    C C H
    D F I
    E E
  `);

  it('renders an initial view with the right panel', () => {
    const store = storeWithProfile(profile);
    // dive() will shallow-render the wrapped component
    const view = shallowWithStore(<Details />, store);
    expect(view.dive()).toMatchSnapshot();
  });

  tabSlugs.forEach((tabSlug: TabSlug) => {
    it(`renders an initial view with the right panel for tab ${tabSlug}`, () => {
      const store = storeWithProfile(profile);
      store.dispatch(changeSelectedTab(tabSlug));

      const view = shallowWithStore(<Details />, store);
      expect(view.dive()).toMatchSnapshot();
    });
  });

  it('show the correct state for the sidebar open button', function() {
    const store = storeWithProfile(profile);
    const view = shallowWithStore(<Details />, store);
    expect(view.dive()).toMatchSnapshot();
    store.dispatch(changeSidebarOpenState('calltree', true));
    expect(view.dive()).toMatchSnapshot();
    store.dispatch(changeSelectedTab('flame-graph'));
    expect(view.dive()).toMatchSnapshot();
    store.dispatch(changeSidebarOpenState('calltree', false));
    store.dispatch(changeSelectedTab('calltree'));
    expect(view.dive()).toMatchSnapshot();
  });
});
