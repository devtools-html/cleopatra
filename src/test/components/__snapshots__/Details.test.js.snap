// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app/Details renders an initial view with the right panel 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="calltree"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details renders an initial view with the right panel for tab calltree 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="calltree"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details renders an initial view with the right panel for tab flame-graph 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="flame-graph"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <FlameGraphView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details renders an initial view with the right panel for tab marker-chart 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={false}
    onSelectTab={[Function]}
    selectedTabSlug="marker-chart"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <Connect(MarkerChart) />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details renders an initial view with the right panel for tab marker-table 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="marker-table"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <Connect(MarkerTable) />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details renders an initial view with the right panel for tab network-chart 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={false}
    onSelectTab={[Function]}
    selectedTabSlug="network-chart"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <WithSizeWrapper />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details renders an initial view with the right panel for tab stack-chart 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={false}
    onSelectTab={[Function]}
    selectedTabSlug="stack-chart"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <Connect(StackChartGraph) />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details show the correct state for the sidebar open button 1`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="calltree"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details show the correct state for the sidebar open button 2`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isopen"
        onClick={[Function]}
        title="Close the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="calltree"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details show the correct state for the sidebar open button 3`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="flame-graph"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <FlameGraphView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;

exports[`app/Details show the correct state for the sidebar open button 4`] = `
<div
  className="Details"
>
  <TabBar
    extraElements={
      <button
        className="sidebar-open-close-button photon-button photon-button-ghost sidebar-open-close-button-isclosed"
        onClick={[Function]}
        title="Open the sidebar"
        type="button"
      />
    }
    onSelectTab={[Function]}
    selectedTabSlug="calltree"
    visibleTabs={
      Array [
        "calltree",
        "flame-graph",
        "stack-chart",
        "marker-chart",
        "marker-table",
      ]
    }
  />
  <ProfileCallTreeView />
  <Connect(CallNodeContextMenu) />
  <Connect(MarkersContextMenu) />
  <Connect(TimelineTrackContextMenu) />
</div>
`;
